include/rocksdb/iterator.h

```diff
@@ -117,6 +117,42 @@ class Iterator : public Cleanable {
     assert(false);
     return Slice();
   }
+
+  /**
+   * true if this iterator is already
+   * checking ReadOptions::iterate_lower_bound,
+   * false otherwise.
+   */
+  virtual bool ChecksLowerBound() const { return false; }
```

Comment:
> Still wow. Baking into a API a way to say "I have a bug and here's what it is. If you wrap me, can you please fix it?"

That's not what it does at all.

If an Iterator which is wrapped supports upper and lower bounds checks, it can inform the wrapping iterator so that the wrapping iterator need not worry about it. It won't cause any harm if they both make the checks. In the past the wrapping iterator had no way of knowing whether the underlying iterator was also making the checks, so it would always have to make the checks itself as well.

Reply:
I still don't understand the use case for exposing what your lower bound is without checking it. If that's a transitional state in migrating code, is it really necessary as part of a public API? Actually, should these functions be protected instead?

Reply:
Nor do I think an iterator needs to be able to express "I would be checking bounds if I had one."

My contention is still that we only need optional upper and lower bounds. If an iterator is simply the UNION of several other iterators, its lower/upper bound would be the smallest/largest (resp.) bound of its constituents, where no bound is considered smaller/larger than any specific value. We don't have to do any addition checking to represent that UNION.

If an iterator is INTERSECTED with a range, we can ignore that additional constraint if our existing lower and upper bounds are within that range. (Performance optimization.) But if our lower/upper bound is outside of that range constraint, we have to use the lower and/or upper range bound as our refined lower/upper bound AND check that bound ourselves.

Perhaps it is helpful to think of the union and intersection roles of a wrapping iterator independently as I have described.

Reply:
@pdillinger I think I agree with your sentiment here and personally am not a fan of these functions -- partially because they are not implemented by more of the Iterator hierarchy. 

Having said that, it seems like there should be a means to tell if an Iterator should check its bounds or assume the parent iterator can do it instead.  I am not sure if these functions are the proper ones to fulfill that contract.  I think the Iterator paradigm should have some sort of "Filter" to allow filtering of bounds (and prefixes).  This might allow some better optimizations as well.  But I am not sure that a "Filter" is in scope for this PR and think it would require more thought to implement properly.

Also, shouldn't the math you stated above be the largest lower bound and smallest upper?

Reply:
I agree with @pdillinger and @mrambacher . The interface of iterator should be made simpler, rather than more complicated. The interface is already in the confusing part, so we should not make it even more confusing. We do have some quite tricky functions in InternalIterator for performance reasons. They are not ideal but having an internal implementation a little bit messier is relatively much more acceptable than make public interface complicated.